File: 010.rb
# 010
# Regular Expression Matching
# 2015-04-18
#####################################################
# @param {String} s
# @param {String} p
# @return {Boolean}
def is_match(s, p)
  l1 = s.size
  l2 = p.size

  i = 0
  j = 0
  while i <= l1 and j < l2
    if p[j+1] == '*'
      if s[i] != p[j] and p[j] != '.'
        j+=2
      elsif is_match(s[i..-1], p[j+2..-1])
        return true
      elsif s[i] == p[j] or p[j] == '.'
        i+=1
      else
        return false
      end
    elsif s[i] == p[j] or p[j] == '.'
      i+=1
      j+=1
    else
      break;
    end
  end

  i == l1 and j == l2
end

#####################################################
# @param {String} s
# @param {String} p
# @return {Boolean}

# DP
def is_match2(s, p)
  l1 = s.size
  m = Array.new(l1+1) {false}
  m[0] = true
  0.upto(p.size-1) do |j|
    if p[j+1] == '*'
      next
    elsif p[j] == '*'
      c = p[j-1]
      1.upto(l1) do |i|
        m[i] = (m[i] or (m[i-1] and (s[i-1] == c or c == '.')))
      end
    else
      c = p[j]
      l1.downto(1) do |i|
        m[i] = (m[i-1] and (s[i-1] == c or c == '.'))
      end
      m[0] = false
    end
  end
  m[l1]
end

#####################################################
s = "aaa";
p = "a*a";
#s = "ab";
#p = ".*..c*";
#s = "baccbbcbcacacbbc";
#p = "c*.*b*c*ba*b*b*.a*";

puts s
puts p
puts is_match(s, p)

File: cpp/001_100/010.cpp
//
//  010.cpp
//  leetcode
//
//  Created by Hang Zhang on 4/15/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <iostream>
using namespace std;

class Solution {
public:
    bool isMatch(string s, string p) {
        int l1 = (int)s.length();
        int l2 = (int)p.length();
        
        int i = 0, j = 0;
        while (i <= l1 && j < l2) {
            if (p[j+1] == '*') {
                if (s[i] != p[j] && p[j] != '.') {
                    j += 2;
                } else if (isMatch(s.substr(i), p.substr(j+2))) {
                    return true;
                } else if (s[i] == p[j] || p[j] == '.') {
                    i++;
                } else {
                    return false;
                }
            } else if (s[i] == p[j] || p[j] == '.') {
                i++;
                j++;
            } else {
                return false;
            }
        }
        return i == l1 && j == l2;
    }

    bool isMatch2(string s, string p) {
        int l1 = (int)s.length();
        int l2 = (int)p.length();

        bool *m = new bool[l1+1]();
        m[0] = true;

        for (int j = 0; j < l2; j++) {
            if (j+1 < l2 && p[j+1] == '*') {
                continue;
            } else if (p[j] == '*') {
                char c = p[j-1];
                for (int i = 1; i <= l1; i++) {
                    m[i] = m[i] || (m[i-1] && (s[i-1] == c || c == '.'));
                }
            } else {
                char c = p[j];
                for (int i = l1; i > 0; i--) {
                    m[i] = m[i-1] && (s[i-1] == c || c == '.');
                }
                m[0] = false;
            }
        }
        bool r = m[l1];
        delete[] m;
        return r;
    }
};

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 010 Regular Expression Matching, C++ ... ...\n";
    /*
    isMatch("aa","a") → false
    isMatch("aa","aa") → true
    isMatch("aaa","aa") → false
    isMatch("aa", "a*") → true
    isMatch("aa", ".*") → true
    isMatch("ab", ".*") → true
    isMatch("aab", "c*a*b") → true
     */
    string s = "";
    string p = "";
    s = "aaa";
    p = "a*a";
    s = "ab";
    p = ".*..c*";
    s = "baccbbcbcacacbbc";
    p = "c*.*b*c*ba*b*b*.a*";

    Solution sol;
    cout << s << ' ' << p << ' ' << (sol.isMatch2(s, p) ? "true" : "false") << '\n';
    
    return 0;
}